README.rst
Tea and Sympathy
::::::::::::::::

This is a technical demonstration of `Balladeer`_.

Balladeer helps you build parser-based Interactive Fiction for the Web.
Balladeer is a Python library.

Installation
++++++++++++

* Windows_
* Linux_

Windows
=======

`Tea and Sympathy` is a command line program.
You use it from the Windows command interpreter.

To launch a new command window:

#. Tap the Windows key so that the Start Menu pops up.
#. Type the word `cmd`.
#. When you see the *Command Prompt* app highlighted, tap the Enter key.

You should see a prompt like this (your user name will differ)::

    Microsoft Windows [Version 10.0.18362.1139]
    (c) 2019 Microsoft Corporation. All rights reserved.

    C:\Users\author>

Prerequisites
-------------

Download and install Python from https://www.python.org/ . You need Python version 3.8 or higher.
Make sure to check the option to add `python` to your environment path.
This makes command line operation more easy.

After you've installed Python, open a command window and type `python`.
You should see something like this::

    C:\Users\author>python
    Python 3.8.2 (tags/v3.8.2:7b3ab59, Feb 25 2020, 22:45:29) [MSC v.1916 32 bit (Intel)] on win32
    Type "help", "copyright", "credits" or "license" for more information.
    >>>

Type `quit()` and press Return.

Virtual Environment
-------------------

#. First make a fresh Python virtual environment::

    python -m venv C:\Users\author\balladeer-app

#. Update the package manager within it::

    C:\Users\author\balladeer-app\Scripts\pip install -U pip wheel

#. Install dependencies::

    C:\Users\author\balladeer-app\Scripts\pip install aiohttp

#. Install (or update) Balladeer::

    C:\Users\author\balladeer-app\Scripts\pip install -U balladeer

Download
--------

#. Download the `repository as a zip file <https://github.com/tundish/tea_and_sympathy/archive/master.zip>`_.
   Unzip it to a local directory.

#. `cd` to `tea_and_sympathy`.

Run
---

You can run the demo in two modes.

#. Launch a local web server to play in a browser (`http://localhost:8080`)::

    C:\Users\author\balladeer-app\Scripts\python -m tas.server

#. Or text-only in the terminal::

    C:\Users\author\balladeer-app\Scripts\python -m tas.story

Linux
=====

The Linux command line is generally more easy to work with than the Windows command prompt.
If you're finding the Windows command prompt tricky, you can install `Git Bash`_ which behaves in a
Linux-like way.

I encourage you to move to a Linux operating system when you are able. 
You can try one out at little cost on a `Raspberry Pi`_ or similar device.

Here are the install instructions for Linux. You need Python version 3.8 or higher.

Virtual Environment
-------------------

#. First make a fresh Python virtual environment::

    python3 -m venv ~/balladeer-app

#. Update the package manager within it::

    ~/balladeer-app/bin/pip install -U pip wheel

#. Install dependencies::

    ~/balladeer-app/bin/pip install aiohttp

#. Install (or update) Balladeer::

    ~/balladeer-app/bin/pip install -U balladeer

Download
--------

#. Download the `repository as a zip file <https://github.com/tundish/tea_and_sympathy/archive/master.zip>`_.
   Unzip it to a local directory.

#. `cd` to `tea_and_sympathy`.

Run
---

You can run the demo in two modes.

#. Launch a local web server to play the web app (`http://localhost:8080`)::

    ~/balladeer-app/bin/python -m tas.app

#. Or text-only in the terminal::

    ~/balladeer-app/bin/python -m tas.story


The freedom to copy
+++++++++++++++++++

You are free to use this project as a teaching example, or as the basis of your own work.
Please read the licence and make sure you `understand the Affero GPL`_.

.. _Balladeer: https://github.com/tundish/balladeer
.. _JSON Feed: https://jsonfeed.org/version/1.1
.. _web rings: https://www.mic.com/p/how-geocities-webrings-made-the-90s-internet-a-cozier-place-19638198
.. _web feeds: https://en.wikipedia.org/wiki/Web_feed
.. _Git Bash: https://gitforwindows.org/
.. _reStructuredText: https://docutils.sourceforge.io/rst.html
.. _Raspberry Pi: https://www.raspberrypi.org/
.. _understand the Affero GPL: https://www.gnu.org/licenses/why-affero-gpl.html



CHANGELOG.rst
..  Titling
    ##++::==~~--''``

.. This is a reStructuredText file.

Change Log
::::::::::

0.11.0
=====

* Add __main__.py for zipapp operation.

0.10.0
=====

* Align with balladeer 0.11.0.

0.9.0
=====

* Align with catchphrase 0.17.0.

0.8.0
=====

* Restore refusal on unmatched command.

0.7.0
=====

* Align with catchphrase 0.15.0.

0.6.0
=====

* Explicitly add drama ensemble from build method.
* Add prompt attribute to drama object.
* Other fixes to align with catchphrase 0.14.0.

0.5.0
=====

* Added a `history` commmand to work as a `transcript <https://github.com/tundish/tea-and-sympathy/issues/2>`_.
* Static pages have a scrollbar.
* `Better reporting on object inspection <https://github.com/tundish/tea-and-sympathy/issues/5>`_.
* More tests and an `improved object model <https://github.com/tundish/tea-and-sympathy/issues/6>`_.
* Fix a bug with pouring hot water.
* Aligned with turberfield-catchphrase 0.13.0 .

0.4.0
=====

* `Improved options for controlling reveal animations <https://github.com/tundish/tea-and-sympathy/issues/3>`_.
* `Differentiated dialogue from game progess <https://github.com/tundish/tea-and-sympathy/issues/1>`_.
* `Improved help messages <https://github.com/tundish/tea-and-sympathy/issues/4>`_.
* `Better documentation of Python install <https://github.com/tundish/tea-and-sympathy/issues/7>`_.

0.3.0
=====

* First public release.

0.2.0
=====

* Added outcomes.
* Aligned with turberfield-catchphrase 0.10.0 .
* Lots of refactors and fixes.

0.1.0
=====

* First running code.



tas/server.py
#!/usr/bin/env python3
# encoding: utf-8

# This is a technical demo and teaching example for the Balladeer library.
# Copyright (C) 2021 D E Haynes

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


import argparse
from collections import deque
from collections import namedtuple
import datetime
import re
import sys
import uuid

from aiohttp import web
import pkg_resources

from balladeer import Presenter

import tas
from tas.story import TeaAndSympathy
from tas.types import Operation


VALIDATION = {
    "email": re.compile(
        "[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]"
        "+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\\.[a-zA-Z0-9]"
        "(?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+"
        # http://www.w3.org/TR/html5/forms.html#valid-e-mail-address
    ),
    "name": re.compile("[A-Z a-z]{2,42}"),
    "session": re.compile("[0-9a-f]{32}"),
}

Session = namedtuple("Session", ["ts", "story"])


async def get_root(request):
    story = TeaAndSympathy()
    story.presenter = story.represent("", facts=story.context.facts)
    session = Session(datetime.datetime.utcnow(), story)
    request.app["sessions"][story.id] = session
    raise web.HTTPFound("/{0.id.hex}".format(story))


async def get_session(request):
    uid = uuid.UUID(hex=request.match_info["session"])
    session = request.app["sessions"][uid]
    story = session.story

    while story.presenter.frames:
        frame = story.presenter.frames.pop(0)
        animation = story.presenter.animate(
            frame,
            dwell=story.presenter.dwell,
            pause=story.presenter.pause
        )
        if animation:
            story.animation = animation
            break

    if story.context.get_state(Operation) == Operation.frames:
        refresh = Presenter.refresh_animations(story.animation, min_val=2)
        refresh_target = story.refresh_target("/{0.hex}".format(uid))
    else:
        refresh = None
        refresh_target = None

    title = next(iter(story.presenter.metadata.get("project", ["Tea and Sympathy"])), "Tea and Sympathy")
    controls = [
        "\n".join(story.render_action_form(action, autofocus=not n))
        for n, action in enumerate(story.actions)
        if story.context.get_state(Operation) not in (Operation.finish, Operation.frames)
    ]
    rv = story.render_body_html(title=title, next_=refresh_target, refresh=refresh).format(
        '<link rel="stylesheet" href="/css/theme/tas.css" />',
        story.render_dict_to_css(vars(story.settings)),
        story.render_animated_frame_to_html(story.animation, controls)
    )

    return web.Response(text=rv, content_type="text/html")


async def post_command(request):
    uid = uuid.UUID(hex=request.match_info["session"])
    session = request.app["sessions"][uid]
    story = session.story
    data = await request.post()
    cmd = data["cmd"]
    if cmd and not story.context.validator.match(cmd):
        raise web.HTTPUnauthorized(reason="User sent invalid command.")
    else:
        text = story.context.deliver(cmd, presenter=story.presenter)
        story.presenter = story.represent(text, facts=story.context.facts, previous=story.presenter)
    raise web.HTTPFound("/{0.hex}".format(uid))


def build_app(args):
    app = web.Application()
    app.add_routes([
        web.get("/", get_root),
        web.get("/{{session:{0}}}".format(VALIDATION["session"].pattern), get_session),
        web.post("/{{session:{0}}}/cmd/".format(VALIDATION["session"].pattern), post_command),
    ])
    app.router.add_static(
        "/css/base/",
        pkg_resources.resource_filename("turberfield.catchphrase", "css")
    )
    app.router.add_static(
        "/css/theme/",
        pkg_resources.resource_filename("tas", "css")
    )
    app["sessions"] = {}
    return app


def main(args):
    app = build_app(args)
    return web.run_app(app, host=args.host, port=args.port)


def parser(description=__doc__):
    rv = argparse.ArgumentParser(description)
    rv.add_argument(
        "--version", action="store_true", default=False,
        help="Print the current version number.")
    rv.add_argument(
        "--host", default="127.0.0.1",
        help="Set an interface on which to serve."
    )
    rv.add_argument(
        "--port", default=8080, type=int,
        help="Set a port on which to serve."
    )
    return rv


def run():
    p = parser()
    args = p.parse_args()

    rv = 0
    if args.version:
        sys.stdout.write(tas.__version__)
        sys.stdout.write("\n")
    else:
        rv = main(args)

    if rv == 2:
        sys.stderr.write("\n Missing command.\n\n")
        p.print_help()

    sys.exit(rv)


if __name__ == "__main__":
    run()
